ouroboros-network/src/Ouroboros/Network/PeerSelection/Churn.hs

```diff
@@ -61,180 +77,363 @@ peerChurnGovernor tracer deadlineChurnInterval bulkChurnInterval psOverallTimeou
-                       min 1 (targetNumberOfActiveBigLedgerPeers base)
-        })
-
-    decreaseEstablishedBigLedgerPeers :: STM m ()
-    decreaseEstablishedBigLedgerPeers =
-        modifyTVar peerSelectionVar (\targets -> targets {
-            targetNumberOfEstablishedBigLedgerPeers =
-              decrease (targetNumberOfEstablishedBigLedgerPeers base -
-                        targetNumberOfActiveBigLedgerPeers base)
-              + targetNumberOfActiveBigLedgerPeers base
-        })
-
-
-
-    go :: StdGen -> m Void
-    go !rng = do
+    increaseActivePeers :: ChurnMode
+                        -> ModifyPeerSelectionTargets
+    increaseActivePeers
+      mode targets
+      =
+      targets {
+        targetNumberOfActivePeers =
+          case mode of
+            ChurnModeNormal  ->
+              targetNumberOfActivePeers base
+            ChurnModeBulkSync ->
+              min 2 (targetNumberOfActivePeers base)
+      }
+
+    checkActivePeersIncreased :: CheckPeerSelectionCounters
+    checkActivePeersIncreased
+      PeerSelectionCounters { numberOfActivePeers }
+      PeerSelectionTargets { targetNumberOfActivePeers }
+      =
+      numberOfActivePeers >= targetNumberOfActivePeers
+
+
+    decreaseActivePeers :: ChurnMode -> ModifyPeerSelectionTargets
+    decreaseActivePeers mode targets =
+      targets {
+        targetNumberOfActivePeers =
+          case mode of
+            ChurnModeNormal ->
+              decrease $ targetNumberOfActivePeers base
+            ChurnModeBulkSync ->
+              min 1 (targetNumberOfActivePeers base - 1)
+      }
+
+    checkActivePeersDecreased :: CheckPeerSelectionCounters
+    checkActivePeersDecreased
+      PeerSelectionCounters { numberOfActivePeers }
+      PeerSelectionTargets { targetNumberOfActivePeers }
+      =
+      numberOfActivePeers <= targetNumberOfActivePeers
+
+
+    increaseEstablishedPeers
+      :: ChurnMode -> UseBootstrapPeers
+      -> ModifyPeerSelectionTargets
+    increaseEstablishedPeers
+      mode ubp targets
+      =
+      targets {
+        targetNumberOfEstablishedPeers =
+          case (mode, ubp) of
+            (ChurnModeBulkSync, UseBootstrapPeers _) ->
+              min (targetNumberOfActivePeers targets + 1)
+                  (targetNumberOfEstablishedPeers base)
+            _  -> targetNumberOfEstablishedPeers base
+      }
+
+    checkEstablishedPeersIncreased :: CheckPeerSelectionCounters
+    checkEstablishedPeersIncreased
+      PeerSelectionCounters { numberOfEstablishedPeers }
+      PeerSelectionTargets { targetNumberOfEstablishedPeers }
+      =
+      numberOfEstablishedPeers >= targetNumberOfEstablishedPeers
+
+
+    increaseEstablishedBigLedgerPeers
+      :: ModifyPeerSelectionTargets
+    increaseEstablishedBigLedgerPeers
+      targets
+      =
+      targets { targetNumberOfEstablishedBigLedgerPeers = targetNumberOfEstablishedBigLedgerPeers base }
+
+    checkEstablishedBigLedgerPeersIncreased
```

Comment:
see `checkEstablishedPeersIncreased` comment.

Reply:
Done.

Reply:
Yeah, it was meant for `checkEstablishedBigLedgerpersDecreased`.